dissect/cstruct/types/structure.py

```diff
@@ -456,9 +503,90 @@ def _write(cls, stream: BinaryIO, data: Union) -> int:
 class Union(Structure, metaclass=UnionMetaType):
     """Base class for cstruct union type classes."""
 
+    _buf: bytes
+
     def __eq__(self, other: Any) -> bool:
         return self.__class__ is other.__class__ and bytes(self) == bytes(other)
 
+    def __setattr__(self, attr: str, value: Any) -> None:
+        if self.__class__.dynamic:
+            raise NotImplementedError("Modifying a dynamic union is not yet supported")
+
+        super().__setattr__(attr, value)
+        self._rebuild(attr)
+
+    def _rebuild(self, attr: str) -> None:
+        if (cur_buf := getattr(self, "_buf", None)) is None:
+            cur_buf = b"\x00" * self.__class__.size
+
+        buf = io.BytesIO(cur_buf)
+        field = self.__class__.fields[attr]
+        if field.offset:
+            buf.seek(field.offset)
+        field.type._write(buf, getattr(self, attr))
+
+        object.__setattr__(self, "_buf", buf.getvalue())
+        self._update()
+
+    def _update(self) -> None:
+        result, sizes = self.__class__._read_fields(io.BytesIO(self._buf))
+        self.__dict__.update(result)
+        object.__setattr__(self, "_values", result)
+        object.__setattr__(self, "_sizes", sizes)
+
+    def _proxify(self) -> UnionProxy:
+        def _proxy_structure(value: Structure) -> UnionProxy:
+            for field in value.__class__.__fields__:
+                if issubclass(field.type, Structure):
+                    nested_value = getattr(value, field.name)
+                    proxy = UnionProxy(self, field.name, nested_value)
+                    object.__setattr__(value, field.name, proxy)
+                    _proxy_structure(nested_value)
+
+        _proxy_structure(self)
```

Comment:
There are no `return`s here, but I guess they're not needed, so the return types should become `None`.